src/http_server.cc

```diff
@@ -3060,120 +3104,449 @@ HTTPAPIServer::HandleInfer(
+    std::map<std::string, triton::common::TritonJson::Value>& input_metadata,
+    const MappingSchema* schema,
+    triton::common::TritonJson::Value& generate_request)
 {
-  HTTPAPIServer::InferRequestClass* infer_request =
-      reinterpret_cast<HTTPAPIServer::InferRequestClass*>(arg);
-
-  evhtp_request_t* request = infer_request->EvHtpRequest();
-  evhtp_send_reply(request, EVHTP_RES_BADREQ);
-  evhtp_request_resume(request);
-
-#ifdef TRITON_ENABLE_TRACING
-  if (infer_request->trace_ != nullptr) {
-    infer_request->trace_->CaptureTimestamp(
-        "HTTP_SEND_START", request->send_start_ns);
-    infer_request->trace_->CaptureTimestamp(
-        "HTTP_SEND_END", request->send_end_ns);
+  // First find all top-level keys in JSON
+  std::vector<std::string> members;
+  RETURN_IF_ERR(generate_request.Members(&members));
+
+  for (const auto& m : members) {
+    auto it = schema->children_.find(m);
+    if (it != schema->children_.end()) {
+      switch (it->second->kind_) {
+        case MappingSchema::Kind::EXACT_MAPPING: {
+          // Read meta data
+          RETURN_IF_ERR(ExactMappingInput(m, generate_request, input_metadata));
+          break;
+        }
+        case MappingSchema::Kind::MAPPING_SCHEMA: {
+          // The key is nested schema
+          triton::common::TritonJson::Value nested_generate_request;
+          RETURN_IF_ERR(generate_request.MemberAsObject(
+              m.c_str(), &nested_generate_request));
+          RETURN_IF_ERR(ConvertGenerateRequest(
+              input_metadata, it->second.get(), nested_generate_request));
+          break;
+        }
+        default:
+          return TRITONSERVER_ErrorNew(
+              TRITONSERVER_ERROR_UNSUPPORTED, "Unsupported schema kind");
+      }
+    } else if (schema->allow_unspecified_) {
+      // Unspecified key follows EXACT_MAPPING
+      RETURN_IF_ERR(ExactMappingInput(m, generate_request, input_metadata));
+    } else {
+      return TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_UNSUPPORTED,
+          "The schema disallow unspecified key");
+    }
+  }
+  return nullptr;  // success
+}
+
+TRITONSERVER_Error*
+HTTPAPIServer::GenerateRequestClass::ExactMappingInput(
+    const std::string& name,
+    triton::common::TritonJson::Value& generate_request,
+    std::map<std::string, triton::common::TritonJson::Value>& input_metadata)
+{
+  auto it = input_metadata.find(name);
+  if (it == input_metadata.end()) {
+    RETURN_IF_ERR(SetTritonParameterFromJsonParameter(
+        name, generate_request, triton_request_));
+  } else {
+    // Parse data type and shape
+    std::string value;
+    it->second.MemberAsString("datatype", &value);
+    auto dtype = TRITONSERVER_StringToDataType(value.c_str());
+
+    // Perform shape validation, assume the value must be either
+    // primitive type or 1-D array.
+    triton::common::TritonJson::Value tensor_data;
+    if (!generate_request.Find(name.c_str(), &tensor_data)) {
```

Comment:
Question / Future:

Could `Members()` return a map of name to JSON Value - and `ExactMappingInput()` take name and value? Just to avoid having to check again - seems we already know its a member (if I understand)

Reply:
Yes. Can you create a ticket for that? I don't want to change the json library in this PR.

Reply:
nvm - thought members was a method of a new class not part of JSON libraryt - agree - not in scope

Reply:
@nnshah1 if you file a ticket, an enhancement I would also like to see is to include the JSON key/name on conversions somewhere. Maybe `Value` can have a `name_` field or something.

Currently if you pass a bad datatype like where a string is expected for the value:
```
{"prompt": 123}
```
the error message would be something like 
```
"attempt to access JSON non-string as string"
```

But it would be a lot easier to debug if the error was:
```
"attempt to access JSON non-string as string for 'prompt' field"
```
or
```
"attempt to access JSON non-string (123) as string"
```
so at least you can associate which field was wrong.

ref: https://github.com/triton-inference-server/common/blob/cf617c93c75d8e03da962115a2471e4b89062aae/include/triton/common/triton_json.h#L713

Reply:
@GuanLuo - is this feasible? I think we do need to have the field name in error message.

Reply:
Yes for `MemberAsXXX` functions, note that it is also change in json library.